### ISSUE TITLE
Tracking Issue for assert_matches

### ISSUE THREAD

**Issue by m-ou-se** [rust-lang/rust#82775]
Feature gate: `#![feature(assert_matches)]`

This is a tracking issue for the `assert_matches!()` and `debug_assert_matches!()` macros.

### Public API

```rust
// core

macro_rules! assert_matches { .. }
macro_rules! debug_assert_matches { .. }
```

### Steps / History

- [x] Implementation: #82770
- [x] Temporarily move the macro into `std::assert_matches::assert_matches` to avoid breakage: https://github.com/rust-lang/rust/issues/82913
- [x] Final commenting period (FCP)
- [ ] Stabilization PR
   - [x] Update the `must_use` message on `is_none` (per https://github.com/rust-lang/rust/pull/62431#pullrequestreview-258576936)
   - [ ] Mention this macro in the documentation of `matches!()`

### Unresolved Questions

- [x] Add `=> expr` syntax?
  - Nope, too confusing. (See discussion below.)

**Comment by m-ou-se** [comment]
Several implementations of `assert_matches!()` in the ecosystem also accept an `=> expr` syntax, to do something with the matched values or even return something from the macro.

- https://docs.rs/assert_matches/1.5.0/assert_matches/macro.assert_matches.html
- https://docs.rs/cool_asserts/1.0.3/cool_asserts/macro.assert_matches.html

This is currently not supported by `std::assert_matches!()` as implemented in #82770.

Adding it would allow things like:

```rust
let thing = assert_matches!(thing, Some(x) => x);
```
and
```rust
assert_matches!(thing, Some((x, y)) => {
    assert_eq!(x.a, 10);
    assert_eq!(y.b(), 20);
});
```

**Comment by drahnr** [comment]
As a (almost) daily user of the `assert_matches` crate, I have to admit that without the additional `$pat => $expr` syntax, for most me it would be still impractical to use.

I.e. asserting a certain sequence of messages with a mocked context and assuring the messages have expected content (which by design may not impl `{Partial,}Eq`) I would usually use:

```rust
assert_matches!(rx.next().unwrap(), Msgs::Specific { a, b, c, .. } => { assert_eq!(a, b); })
assert_matches!(rx.next().unwrap(), Msgs::Specific { a, b, c, .. } => { assert_eq!(b, c); })
..
```

so for this the extension with ` => ` is required to superseed `assert_matches` crate in practice from my experience.

**Comment by m-ou-se** [comment]
In cases like that, the `if` guard on the pattern also works, right?
```rust
assert_matches!(rx.next(), Some(Msgs::Specific { a, b: B(b), .. }) if a == b);
```

**Comment by drahnr** [comment]
While your approach is valid for simple cases, I disagree for the general use case. This is not very ergonomic as soon as there are various sub patterns, that have to be asserted against transformed values or larger enums with multiple elements 

```rust
assert_matches!(rx.next(), Some(Msgs::Specific { a, b: B(b), .. }) => {
  assert_eq!(a, b);
  let container = transform($container);
  if container.is_empty() {
       assert!(..);
  } else {
      assert!(..);
  }
});
..

It could still be expressed with boolean expression combinators, yet that adds a lot visual complexity that imho should not exist in a test case.

**Comment by m-ou-se** [comment]
@rust-lang/libs Ping for opinions.

Should `assert_matches` accept `=> { .. }` blocks to allow for nested asserts (and other code)? The existing `assert_matches` macros in the ecosystem accept that. But I personally feel it makes things a bit too complicated and unclear.

**Comment by BurntSushi** [comment]
It's an interesting extension. At first it looked a bit odd to me, but after ruminating on it a bit, it does seem fairly natural to me.

In terms of moving forward:

1. Would adding this later be a backwards compatible addition?
2. How often is this extension used in practice? In theory, we could search the code on crates.io and count the ratio between uses of `assert_matches!` with and without this extension. If the ratio is very small, then maybe we land it without the extension if the answer to (1) is "yes."

**Comment by joshtriplett** [comment]
This makes me think of the [proposal to add `=>` syntax to `matches!`](https://github.com/rust-lang/rust/pull/79188), which the libs team declined.

This seems very similar, and I feel like the arguments are the same. If we have `assert_matches!(x, Some(Variant(a, b)) => a == b)`, similar use cases would motivate `if matches!(x, Some(Variant(a, b)) => a == b) { ... }`.

I personally feel that in both cases the `=>` syntax adds enough complexity to make it worth spelling out a `match` or `if let`, instead. I don't really want to see an entire block of arbitrary code embedded inside an `assert_matches!`. I also think this will become easier to write when we finish `if let` chaining.

**Comment by m-ou-se** [comment]
`assert_matches!(x, Some((a, b)) if a == b)` would assert that x is a Some and a and b are equal, but `assert_matches!(x, Some((a, b)) => a == b)` would only assert that x is a Some, and *return* the value of `a == b` without asserting anything about it. The difference there is far to subtle for me.

We could require the `=> expr` to be `()`, but I don't think that solves the problem of unclear semantics entirely.

**Comment by joshtriplett** [comment]
> `assert_matches!(x, Some((a, b)) if a == b)` would assert that x is a Some and a and b are equal, but `assert_matches!(x, Some((a, b)) => a == b)` would only assert that x is a Some, and _return_ the value of `a == b` without asserting anything about it. The difference there is far to subtle for me.

For me as well; that was not clear to me at all.

**Comment by BoxyUwU** [comment]
A separate macro that always requires a `=> expr` arm i.e. `unwrap_matches` `extract_matches` might make it less ambiguous/more clear

**Comment by gilescope** [comment]
Well if we're agreed on leaving it as is then we can head towards stabilising it...

**Comment by gThorondorsen** [comment]
<details><summary>@drahnr in <a href="https://github.com/rust-lang/rust/issues/82775#issuecomment-804807877">#82775 (comment)</a></summary>

> While your approach is valid for simple cases, I disagree for the general use case. This is not very ergonomic as soon as there are various sub patterns, that have to be asserted against transformed struct member values or larger enums with multiple elements common for messaging enums.
> 
> ```rust
> assert_matches!(rx.next(), Some(Msgs::Specific { a, b: B(b), .. }) => {
>   assert_eq!(a, b);
>   let container = transform($container);
>   if container.is_empty() {
>        assert!(..);
>   } else {
>       assert!(..);
>   }
> });
> ..
> ```
> 
> It could still be expressed with boolean expression combinators using the current impl, yet that adds a lot visual complexity that imho should not exist in a test case.
> 
> Note that I am not saying the impl is bad or anything :) - I would very much like to see `assert_matches!` in `std`! This effort is much appreciated!

</details>

Your example can be rewritten to a match guard that always returns true except when it panics, like this.
```diff
-assert_matches!(rx.next(), Some(Msgs::Specific { a, b: B(b), .. }) => {
+assert_matches!(rx.next(), Some(Msgs::Specific { a, b: B(b), .. }) if ({
     assert_eq!(a, b);
     let container = transform($container);
     if container.is_empty() {
         assert!(/* … */);
     } else {
         assert!(/* … */);
     }
-});
+    true
+}));
```
Definitely non-obvious when you don't know about it, but hardly worse to write if you are adding to an existing test suite with the pattern everywhere.

**Comment by drahnr** [comment]
Re-reading the comments, I fully support the current impl and withdraw my previous concerns.

**Comment by fee1-dead** [comment]
I would prefer the macro to expand as `let .. else { panic!() }` and users would be able to use the bindings after the assertion.

Does this sound reasonable? Would `feature(let_else)` block the stabilization of this macro, or can we use `allow_internal_unstable`?

**Comment by steffahn** [comment]
I think there's a lot of value in having `assert_matches!(...)` be basically just an `assert!(matches!(...))` with better error messages (and hence with an additional `Debug` bound). It makes the macro much easier to understand if those two expressions are essentially equivalent. Even if both `assert_matches` and `matches` would both support the `=> expr` syntax, this rule would be broken: At least following the existing (rejected) proposal, the "equivalent" to `assert_matches!(... => ...)` would suddenly be `matches!(... => ...).unwrap()`.

**Comment by m-ou-se** [comment]
Looks like we reached consensus on not having the `=>` syntax.

@rfcbot merge

**Comment by m-ou-se** [comment]
If we stabilize this as `std::assert_matches!()`, then we'll break a few crates. See https://github.com/rust-lang/rust/issues/82913.

```
chrysanthemum-0.0.1/src/verify.rs:                             use assert_matches::*;
error-context-0.1.2/src/lib.rs:                                use assert_matches::*;
port-finance-lending-0.1.0/tests/helpers/mod.rs:               use assert_matches::*;
rfc6381-codec-0.1.0/src/lib.rs:                                use assert_matches::*;
lethe-0.6.0/src/ui/args.rs:                                    use assert_matches::*;
lethe-0.6.0/src/actions/wipe.rs:                               use assert_matches::*;
rant-4.0.0-alpha.9/tests/runtime_tests.rs:                     use assert_matches::*;
kekbit-core-0.2.3/src/shm.rs:                                  use assert_matches::*;
port-finance-variable-rate-lending-0.1.2/tests/helpers/mod.rs: use assert_matches::*;
```

@rust-lang/libs-api How should we continue?

**Comment by m-ou-se** [comment]
cc'ing the macro resolution expert: @petrochenkov. maybe you have any ideas?

**Comment by yaahc** [comment]
> If we stabilize this as `std::assert_matches!()`, then we'll break a few crates. See #82913.

I'm confused, how are these still being broken after https://github.com/rust-lang/rust/pull/87195?

**Comment by jo-so** [comment]
Would it be possible to allow `=> $x:ident` to make it possible to unwrap with `assert_matches!`? I often have tests like:

```rust
let x = assert_matches!(fun(), Ok(x) => x);
assert_eq!(x.name, "Test");
assert_eq!(x.fld, 12);
```

Now I have to write this what's a little longish.

```rust
let x = fun();
assert_matches!(x, Ok(_));
let x = x.unwrap();
assert_eq!(x.name, "Test");
assert_eq!(x.fld, 12);
```

**Comment by sfackler** [comment]
What's the purpose of using assert_matches at all there?

**Comment by jhpratt** [comment]
That has been previously discussed and decided against.

**Comment by m-ou-se** [comment]
> I'm confused, how are these still being broken by `assert_matches` after #87195?

I thought moving the macro into that module without re-exporting it in the prelude or crate root was only a temporary fix while this feature is unstable. I think once this is stable, it shouldn't require an extra import, to make sure it's available in the same way as `assert_eq!()` etc.

**Comment by jo-so** [comment]
Steven Fackler schrieb am Wed 16. Mar, 05:02 (-0700):
> What's the purpose of using assert_matches at all there?

I want to assert that the return type matches `Ok()` and want to fail
otherwise. If it's Ok(), I want to go on and run other checks.

**Comment by jo-so** [comment]
> How would the behavior of your test change if you removed that line?

It wouldn't compile, because the variable is not defined.

**Comment by sfackler** [comment]
`assert_matches!(x, Ok(_));` does not define any variables.

**Comment by Shatur** [comment]
> Now I have to write this what's a little longish.

I would use `expect()` in your case:

```rust
let x = fun().expect("Should be ok");
assert_eq!(x.name, "Test");
assert_eq!(x.fld, 12);
```

**Comment by TennyZhuang** [comment]
> Would it be possible to allow `=> $x:ident` to make it possible to unwrap with `assert_matches!`? I often have tests like:
> 
> ```rust
> let x = assert_matches!(fun(), Ok(x) => x);
> assert_eq!(x.name, "Test");
> assert_eq!(x.fld, 12);
> ```
> 
> Now I have to write this what's a little longish.
> 
> ```rust
> let x = fun();
> assert_matches!(x, Ok(_));
> let x = x.unwrap();
> assert_eq!(x.name, "Test");
> assert_eq!(x.fld, 12);
> ```

IMO feature let_else can help you in the case.

**Comment by jo-so** [comment]
> `assert_matches!(x, Ok(_));` does not define any variables.

I'm talking about `let x = assert_matches!(fun(), Ok(x) => x);`

**Comment by jo-so** [comment]
> > Now I have to write this what's a little longish.
> 
> I would use `expect()` or `unwrap()` in your case:
> 
> ```rust
> let x = fun().expect("Should be ok or any other explanation");
> assert_eq!(x.name, "Test");
> assert_eq!(x.fld, 12);
> ```

Yes, `expect()` would allow to give a static message. But `assert` has that advantage to see the bad value. Sometimes tests trigger nasty bugs, and it is helpful to have as much information as possible. Hence, I use `assert`.

**Comment by SimonSapin** [comment]
`Result::<T, E>::expect` requires `E: Debug` and prints the error value in the panic message: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c23502bf5661ca1d97b5f691e826e851

**Comment by luqmana** [comment]
Trying to collect together the current state of things:

`assert_matches!` is currently a `decl_macro` defined at `core::macros::assert_matches` but exported as `core::assert_matches::assert_matches`. `libstd` then re-exports it at `std::assert_matches::assert_matches`.

Originally introduced without the submodule (and as a regular `#[macro_export] macro_rules`) in #82770, it unfortunately caused some breakage in the wider ecosystem due to existing code importing similar macros of the same name via globs. Unfortunately, the compiler couldn't resolve the ambiguity between the macro from the prelude and the macro from the glob import, e.g in fuchsia:

```
error[E0659]: `assert_matches` is ambiguous (glob import vs any other name from outer scope during import/macro resolution)
   --> ../../src/lib/test_executor/rust/src/diagnostics.rs:363:13
    |
363 |             assert_matches!(
    |             ^^^^^^^^^^^^^^ ambiguous name
    |
    = note: `assert_matches` could refer to a macro from prelude
note: `assert_matches` could also refer to the macro imported here
   --> ../../src/lib/test_executor/rust/src/diagnostics.rs:322:13
    |
322 |             super::*,
    |             ^^^^^^^^
    = help: consider adding an explicit import of `assert_matches` to disambiguate
    = help: or use `self::assert_matches` to refer to this macro unambiguously
```

This prompted the move to a separate submodule (and the change to a `decl_macro`) in #86947.

As it stands, ideally we'd like to have `assert_matches` closely mirror the other existing assert macros as much as possible but there are some difficulties. The options thus far seem to be:

1. Changing the import behaviour to resolve the ambiguity in favor of the explicit `use` (via glob) over the prelude.
2. (a) Stabilize it as-is in the submodule, and
    (b) later expose it via prelude in a new edition.
3. No explicit handling and just let it get to stable in the prelude.

**Option 1**: may be [tricky](https://github.com/rust-lang/rust/issues/65721#issuecomment-546441859) (cc @petrochenkov on if anything's changed there?).

**Option 2**: (a) introduces some asymmetry compared to `assert_eq!` which is not great. (b) also means adding a migration lint for the next edition to automatically fix future edition-incompatible code. Without an auto-migration, the friction to upgrading editions would be that much higher which isn't something we want to encourage.

**Option 3**: letting it rid through nightly and hoping most of the ecosystem adjusts their code is not a great look. It unfortunately happened when `matches!` landed and ended up [making it to stable](https://github.com/rust-lang/rust/issues/66518#issuecomment-568197058) without being addressed. Not really something we want to repeat (hence #86947 to begin with). And really it's not unexpected we'll run into this again so better to find a solution now. (Though, perhaps it's worth a crater run to see what the impact looks like today?)
 
(All the above also applies to `debug_assert_matches!`)


Finally thanks to everyone (@m-ou-se especially!) for all the work and input thus far.

**Comment by petrochenkov** [comment]
>Option 1: may be https://github.com/rust-lang/rust/issues/65721#issuecomment-546441859 (cc @petrochenkov on if anything's changed there?).

Nothing changed, we still cannot remove the error without [proving that with any expansion order and any import resolution order `matches` materializes from the `use foo::*` glob "not later" than from prelude](https://github.com/rust-lang/rust/issues/65721#issuecomment-546441859).

We can take "the `foo::*` glob imports from a different crate" as a proof (assuming we can also prove that the `use foo` prefix is guaranteed to be resolved not later than `assert_matches` from the prelude).
All of this may require an uncertain amount of time and refactorings in rustc_resolve.

It will not help with the `use super::*;` case from the previous comment though, because it's an import from the current crate.

**Comment by nnethercote** [comment]
I'm very late to the party, but is `assert_not_matches` worth considering? It came up in [a tweet](https://twitter.com/nnethercote/status/1546673539214118912):
> This Rust code is very shouty:
>
> `assert!(!matches!(x, A));`

That tweet was inspired by some real code I wrote today.

**Comment by nox** [comment]
I think `assert_matches` should have a `#[deny(unused_variables)]` on the success arm.

I found this at work:

```rust
let metadata = some_operation_under_test();
let expected_metadata = some_expected_value();

assert_matches!(metadata, expected_metadata);
```

This emitted two warnings about `expected_metadata` being unused, but nobody noticed and the commit landed. I fixed the issue (by using `assert_eq`), but it turns out the test don't actually pass hah.

If there was `#[deny(unused_variables)]`, it would not compile at all and the problem would have been detected earlier.

What is the benefit of allowing variable names in there anyway?

**Comment by SimonSapin** [comment]
I think the ideal would be for `assert_matches!` to warn if it is given any irrefutable pattern

**Comment by nox** [comment]
There is no existing lint for irrefutable patterns in `match` expressions, as they are expected there. What's not expected is the unreachable pattern for the failing path, and that indeed already emits a warning currently.

I still think this should emit an error. I understand that `unused_variables` is a warning in the general case, as you may have forgotten to use the variable, but there is no way ever to use variables from a pattern in `assert_matches`.

**Comment by safinaskar** [comment]
I want to say again what @fee1-dead said in https://github.com/rust-lang/rust/issues/82775#issuecomment-993680387 : let's make `assert_matches!(expr, pat)` expand to `let pat = expr else { panic!(); }`. Note that let-else is already in stable.

Such expansion will make both `assert_matches!(expr, pat if cond)` and `assert_matches!(expr, pat => expr2)` unneeded. And my expansion will be more powerful than both this variants.

Code `let x = assert_matches!(fun(), Ok(x) => x)` mentioned by @jo-so in https://github.com/rust-lang/rust/issues/82775#issuecomment-1069049011 will be written simply as `assert_matches!(fun(), Ok(x))`.

Code `assert_matches!(c, Ok(x) | Err(x) if x.len() < 100)` from official docs ( https://doc.rust-lang.org/nightly/std/assert_matches/macro.assert_matches.html ) will be written so:

```rust
assert_matches!(c, Ok(x) | Err(x)); // Line 1
assert!(x.len() < 100); // Line 2
```

Yes, now the code is longer, but it is more explicit, and in my opinion is more clear. Moreover, now fail at line 1 and fail at line 2 will give two different error messages (this is helpful for debugging), as opposed to current implementation.

You may say that this proposal is too radical. Yes, possible. Okay, then keep `assert_matches!` simple and also introduce `assert_let!` macro based on my idea

**Comment by safinaskar** [comment]
You may say: "Just write `let pat = expr else panic!()`". Yes, but in this case we lose this detailed error message

**Comment by safinaskar** [comment]
@drahnr's example from https://github.com/rust-lang/rust/issues/82775#issuecomment-804807877 will be written so:
```rust
assert_matches!(rx.next(), Some(Msgs::Specific { a, b: B(b), .. }));
assert_eq!(a, b);
let container = transform($container);
if container.is_empty() {
     assert!(..);
} else {
    assert!(..);
}
```
This is more clear than both original version and @gThorondorsen's version from https://github.com/rust-lang/rust/issues/82775#issuecomment-983759924

**Comment by marmeladema** [comment]
I know this is kind of expected because of because of how patterns are working but I feel like `assert_matches` lacks usability:
* You cannot use function calls
* You cannot use `..Default::default()` or similar
* You cannot declare ranges like `1..3` because they are themselves patterns

See https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=289d8072098365fbaea3e59a4f8d185d

My use case what that I had an internal `PartialEq` struct that had a bunch of tests using `assert_eq` macro. I wanted to add a non-`PartialEq` field. Thus my top struct would not implement `PartialEq` anymore, so I figured I could just swap all usage of `assert_eq` to `assert_matches`  and add `not_partial_eq_field: _` everywhere and be done with it but it turns out I was very wrong.

**Comment by fee1-dead** [comment]
I think you are confusing patterns and expressions. `assert_eq` compares two expressions while `assert_matches` ensures that the value matches a pattern (the same you use in `if let`, `while let`, and `match`)

You can't make a pattern that matches "the default fields a struct has" since that would require computing the default value of that type in compile time, requiring const traits, an unstable feature.

To check that a range matches what you want, use: `Range { start: 1, end: 4 }`.

**Comment by marmeladema** [comment]
> To check that a range matches what you want, use: Range { start: 1, end: 4 }

Thanks! Not sure why I haven't thought of that myself :(

> I don't think we have an easy way to mix arbitrary expressions and patterns together. Maintaining a separate grammar for assert_matches just for that feels wrong. assert_matches should do what it says it does.

I fully agree, but at the same time, I feel (and its really a feeling rather than a real argument) like those "restrictions" seem natural when used in match/if-let etc whereas when used in an assert, it felt very weird.

I think my confusion comes from the fact that the pattern syntax is *very* similar to expression syntax (and for ranges, you can use the exact same syntax) but they mean very different things.

**Comment by lo48576** [comment]
I like the idea of destructuring with refutable pattern, but I don't like macros named `assert_*` introducing new bindings. Calling it "assert" is inconsistent and unintuitive. It looks like "unwrap" or "expect", but not "assert".

This would be bikeshed, but I think they should be called by more straight-forward names such as, for example, `bind_match_or_panic!(pat, expr);` or `let_match_expect!(pat, expr);`, **if** we allow them introduce new variables outside the macro call.
`assert` macros should just assert, not introduce new names.

**Comment by jplatte** [comment]
I see what you mean, but I've now published this under the same macro name at first, since I'm not a fan of the two potential alternative names you came up with. I called the crate [assert_matches2](https://crates.io/crates/assert_matches2).

**Comment by jo-so** [comment]
Jonas Platte schrieb am Mi 07. Jun, 16:59 (-0700):
> I see what you mean, but I've now published this under the same macro name
> at first, since I'm not a fan of the two potential alternative names you
> came up with. I called the crate
> [assert_matches2](https://crates.io/crates/assert_matches2).

How about `unwrap_match!()` or `matched_unwrap!()`?

**Comment by jplatte** [comment]
I think personally I'll use it as-is for now, but another option would be `let_else_panic!($pat = $expr $(, $format_args)?)`.

**Comment by jplatte** [comment]
I have now added the second syntactic form, spelled `assert_let!(<pattern> = <expression>);`, to `assert_matches2`. Now that I'm re-reading I see that @safinaskar previously had the same idea for a name :)

Also, the option of stabilizing without being in the prelude was discussed above, and the only reason against it seemed to be that it would require time and effort to then add it to the prelude in a coming edition. But how is this a reason not to stabilize and let people start using the macro with an explicit import? Nobody would be forced to get the automatic migration (or nameres refactor) done before any specific edition. If it doesn't happen, it doesn't happen and the macro stays available as something you can import explicitly. Same as now really, except it doesn't require adding a dependency. I think that's strictly better than the status quo.

**Comment by Fishrock123** [comment]
I think it would be unintuitive to call this anything other than `assert_matches!` (at least for a version which does not make a let binding). I think that both should probably exist, I think it would be strange to most that `assert_matches!` would produce a binding, and I think in a lot of cases (`Result`) the resulting binding may be something that is marked `#[must_use]` (now or in the future), which would be rather undesirable to be forced to always unnecessarily handle.

**Comment by Voultapher** [comment]
I fear it might be a bit too late, but I recently came across `assert_matches` and the `if` part of the feature is *very* confusing. Let's take the provided example:

```rust
assert_matches!(c, Ok(x) | Err(x) if x.len() < 100);
```

What does that code check?

A) The value `c` is allowed to be either `Ok` under all len, and `Err` *if* `x.len() < 100`. What part does the `if` bind to?
B) It's allowed to be either `Ok` or `Err` as long as `x.len() < 100`. No assertion about situations where `x.len() >= 100`.
C) It's allowed to be either `Ok` or `Err` *AND* in both cases `x.len() < 100` has to be true.

I worry that this feature will confuse both existing and especially new users. In Rust and pretty much every other programming language the keyword `if` is associated with conditionality. Here we are adding a feature where `if` has zero conditionality. As a stopgap measure we could stabilize only the expression without `if`.

**Comment by Allen-Webb** [comment]
> I fear it might be a bit too late, but I recently came across `assert_matches` and the `if` part of the feature is _very_ confusing. Let's take the provided example:
> 
> ```rust
> assert_matches!(c, Ok(x) | Err(x) if x.len() < 100);
> ```
> 
> What does that code check?
> 
> A) The value `c` is allowed to be either `Ok` under all len, and `Err` _IF_ `x.len() < 100`. What part does the `if` bind to? B) It's allowed to be either `Ok` or `Err` _IF_ `x.len() < 100`. No assertion about situations where `x.len() >= 100`. C) It's allowed to be either `Ok` or `Err` _AND_ in both cases `x.len() < 100` has to be true.
> 
> It's C. The one without any conditional logic.
> 
> I worry that this feature will confuse both existing and especially new users. In Rust and pretty much every other programming language the keyword `if` is associated with conditionality. Here we are adding a feature where `if` has zero conditionality. As a stopgap measure we could stabilize only the expression without `if`.

I agree this situation is confusing, but it isn't something new. It is the same syntax that shows up in match statements, so I am of the opinion if it gets fixed it should get fixed at a higher level than `assert_matches` and it would be more confusing / disruptive for `assert_matches` to behave differently than `match`.

**Comment by Voultapher** [comment]
I disagree with equating the `if` in `match` statements to this situation. I think they are meaningfully different.

Take this as an example:

```rust
let x: Result<String, u64> = Ok("xlk".into());

match x {
    Ok(val) if val.len() < 5 => println!("A"),
    Err(_err) => println!("B"),
    _ => println!("C"),
}
```

To me the salient differences are:

1. It's much more clear what the if is binding and applying to.
2. If a pattern becomes non-exhaustive, users notice immediately. In contrast `assert_matches` silently swallows wrong assumptions about pattern exhaustiveness. This in itself is not bad, but it compounds the footgun potential.
3. A match is structured in a way that makes it clear that there are different "branches" of execution, and the if is easy to conceptualize as restricting the branch to values that match the condition. Arguably with `assert_matches` you have the same conceptual logic. But because it's inverted, you are not describing "If it is this type or pattern and fulfills these conditions" as you do in `match`, you are describing `If it is not this type or pattern and does not satisfy these conditions, give me an panic`. I think this inversion of logic makes this significantly different to `if` in `match`.

I sympathise with wanting to stick to the same mechanism as `match`, but I think the factors I've described combine to create something that is rather easy to use incorrectly and misunderstand. Simple and consistent rules don't always result in easy to use experiences. One example is the 2018 Rust modules change, that did away with a notionally simpler set of rules, that despite their surface level simplicity tripped up users more often than need be. I'm willing to bet that if you show the example above to new Rust users, and ask them if they think it will be A, B, or C a significant percentage will get it wrong. That's bad for a systems programming language that prides itself in explicit code.

**Comment by Allen-Webb** [comment]
> That even worse alternatives exist, does not justify adding something confusing to the standard library.

I don't believe this logic applies here. From my viewpoint this is `assert_matches!(expression, pattern)`. If we make a one-off change in the behavior of pattern for `assert_matches!`, that to me is more problematic than consistency with other places we handle patterns like match.

I don't think this is a fair example:
```rust
match x {
    Ok(val) if val.len() < 5 => println!("A"),
    Err(_err) => println!("B"),
    _ => println!("C"),
}
```
You need something more like:

```rust
match x {
    Ok(x) | Err(x) if x.len() < 5 => println!("A"),
    _ => println!("B"),
}
```

How is this any less confusing than the following?
```rust
assert_matches!(c, Ok(x) | Err(x) if x.len() < 100);
```

**Comment by Voultapher** [comment]
> How is this any less confusing than the following?

Ok, so they do different things, so it's not a clear comparison. Even then, I think my point still stands. I think it is meaningfully different and more confusing. The `match` example has two arms, `match` is a well known concept in Rust to describe control flow, so is `if`.  `assert_matches` does very little to help users via, indentation, blocks etc. to help users understand the control flow implications. The core difference I argue is, that in the `match` example `if` controls whether or not the branch is taken. In contrast the `assert_matches` has it where, the `if` controls the branch that conceptually is *not* taken. If the condition in the `if` is false then the control flow changes. There is also the language angle. "Match this arm if there is this condition met." that's relatively straight forward to a non native english speaker like myself. The `assert_matches` reads like "Assert that this matches (is true) as long as (if) this condition is met.", I find that confusing and a significant difference. And I'm willing to bet that significantly more people will guess the behavior of the `assert_matches` example wrong, than the `match` example.

Also, please stop assuming I'm suggesting a different syntax or alternatives. All I'm saying is, we have an understandability problem here. I'm not saying how we should fix it, that's the next step. The first step would be acknowledging that `match` and `assert_matches` are different enough that different levels of footguns and understanding issues are possible.

And to those expressing their disagreement with my viewpoints as dislikes, I kindly ask you to formulate your viewpoints and explain why you disagree with mine.

**Comment by ssokolow** [comment]
> And to those expressing their disagreement with my viewpoints as dislikes, I kindly ask you to formulate your viewpoints and explain why you disagree with mine.

For me, my biggest concern is ensuring that, once the nature of pattern syntax *has* been learned, the number of "dialects" of it that must be memorized doesn't grow. (i.e. "refutable" vs. "irrefutable" is already enough in a language with so much other complexity that also needs to be learned.)

**Comment by Voultapher** [comment]
I agree that having multiple pattern syntaxes would be a bad thing. I believe that, in this situation the choice of the word `assert_matches` in combination with the `if` allowed in the `match` arm syntax is a serious footgun, and having multiple syntaxes, can both be true at the same time. I suspect some of the issues can be alleviated by better documentation, something I'm open to contribute to. At the same time, it's not clear to me that reacting to "we have a problem" with refusing the problems's existence by pointing to problematic alternatives is a good course to follow.

**Comment by Allen-Webb** [comment]
> refusing the problems's existence by pointing to problematic alternatives is a good course to follow.

I think you are misunderstanding what i am trying to say. I am saying if you think the if syntax is confusing, we should fix it in the general sense, not for `assert_matches`. I don't see this as the right place to try to address changes to the pattern syntax.

Maybe I am misunderstanding you if you are talking about different comments.

**Comment by Voultapher** [comment]
To me I'm mostly fine with pattern syntax in existing places, though it's difficult to asses ones own biases and knowledge curse. For me `assert_matches` presents a unique form of confusion risk. That might sound surprising, given in the other places like `match` I'm happy with the syntax. The same rules can leave to desirable and undesirable outcomes in different places. I'm not sure how to fix this, the design space is large. For me the first step would be acknowledging the potential problem and doing some experiments that check my assumption of how confusing this really is. Earlier I've asked a college and they were confused by the syntax, as was I. That's sample size 2, we can do better.

Also there is precedence for not allowing the full pattern syntax everywhere, for example `if let Ok(x) if x.len() < 10 = x { ... }` is not valid.

**Comment by tmccombs** [comment]
> Ok, so they do different things, so it's not a clear comparison.

```rust
assert_matches!(c, Ok(x) | Err(x) if x.len() < 100);
```

is basically equivalent to :

```
match {
  Ok(x) | Err(x) if x.len() < 100 => {}
  _ => panic!("a meaningful error message");
}
```

If the `if` guard isn't allowed in `assert_matches` or uses a different syntax it begs the question of why it isn't consistent with the equivalent match expression.

> It's much more clear what the if is binding and applying to.

I don't think this is any less of a problem in the match example above. Although, in my opinion is is pretty clear that `|` binds more tightly than `if`.

> And to those expressing their disagreement with my viewpoints as dislikes, I kindly ask you to formulate your viewpoints and explain why you disagree with mine.

Well, I've mostly just agreed with others have said and haven't had a lot else to add to it. Or time to formulate a more thorough response.



I do have a couple of ideas that might help with this though:

1. Have a lint that suggests adding parenthesis around multiple patterns separated by `|` if there is an if guard (I'm actually not entirely sure if that is valid syntax today, so maybe allow that if it isn't currently allowed). So for example suggest that you use `(Ok(x) | Err(x)) if x.len() < 100` instead of `OK(x) | Err(x) if x.len()`.
2. If we had postfix macros, this might be a good candidate: IMO `c.assert_matches!(Ok(x) | Err(x) if x.len() < 100)` is a little more clear that the if is part of the match pattern instead of saying only match if a condition is true. 
3. Allow `&&` instead of `if`. If we were to stay consistent this would need to be allowed in `match` expressions as well, which would require a separate RFC.  But such a change could be added later.

**Comment by Voultapher** [comment]
> I don't think this is any less of a problem in the match example above. Although, in my opinion is is pretty clear that | binds more tightly than if.

I should clarify, I meant it's more clear it applies to the branch, as highlighted by the separate line and something following it, the branch body. In `assert_matches` it is the end of the expression, making that IMO less clear.

**Comment by Allen-Webb** [comment]
> Have a lint that suggests adding parenthesis around multiple patterns separated by | if there is an if guard (I'm actually not entirely sure if that is valid syntax today, so maybe allow that if it isn't currently allowed). So for example suggest that you use (Ok(x) | Err(x)) if x.len() < 100 instead of OK(x) | Err(x) if x.len().

Something like this could apply in all cases patterns are used in `macro_rules!`, but it isn't valid syntax today. It would also be a bit weird to allow it in that case and not allow cases like `Ok(x) | {Err(x) if x.len() < 100)`

**Comment by nnethercote** [comment]
FWIW, `assert!(matches!(c, Ok(x) | Err(x) if x.len() < 100));` is already valid.

Also, about this running example: binding a variable in multiple arms and then using it in a condition is pretty unusual.

If you don't like the fact that `Ok(x) | Err(x) if x.len() < 100` is a valid pattern, I can understand that. It's arguably surprising. But I don't buy the argument that `assert_matches!` is particularly different to `match` or anywhere else a pattern can appear.

**Comment by nsunderland1** [comment]
I think it's also worth thinking about what the worst case is. If I write this, expecting the `if` to only apply to the `Err` case:

```rust
assert_matches!(c, Ok(x) | Err(x) if x.len() < 100);
```

The result is a _strictly stronger_ assertion than I intended to write. There are cases where that might be undesirable, but it's nowhere near as bad as an unintentionally weak assertion. It will never let invariant violations slip through, the worst case is an unneeded panic.

**Comment by Voultapher** [comment]
> But I don't buy the argument that assert_matches! is particularly different to match or anywhere else a pattern can appear.

I think the difference comes down to english grammar. "match if" makes sense, "assert that pattern if condition" also makes sense, however that's not the semantics. It's either "panic if not pattern and or not condition" or "assert that pattern and condition". Am I really the only one here that finds that problematic?

**Comment by fee1-dead** [comment]
> I think it's also worth thinking about what the worst case is. If I write this, expecting the `if` to only apply to the `Err` case:
> 
> ```rust
> assert_matches!(c, Ok(x) | Err(x) if x.len() < 100);
> ```
> 
> The result is a _strictly stronger_ assertion than I intended to write.

Perhaps I can assume that if you only wanted the pattern to apply to the `Err` case, you would probably write:

```rust
assert_matches!(c, Ok(_) | Err(x) if x.len() < 100);
```

and get an error:
```
error[E0408]: variable `x` is not bound in all patterns
 --> src/main.rs:4:24
  |
4 |     assert_matches!(c, Ok(_) | Err(x) if x.len() < 100);
  |                        ^^^^^       - variable not in all patterns
  |                        |
  |                        pattern doesn't bind `x`
```

which *could* give you some clue that you can't really do this. (if this confusion is common, then it would be a diagnostics issue, and we can suggest removing the `Ok(_) |` altogether)

For people who aren't familiar with `matches!`, though, the following statement would probably be equally confusing:

```rust
if matches!(c, Ok(x) | Err(x) if x.len() < 100) {
        
}
```

and if you tried to read it out loud token by token it would sound like "if matches c Ok(x) or Err(x) if the length of x is smaller than 100".

I think if we were to stabilize `assert_matches!` the syntax should be equivalent to `assert!(matches!`. An alternative syntax like `matches!(Ok(x) | Err(x) && x.len() < 100)` could be equally (or actually more, IMO) confusing as to why it doesn't follow the same if guard syntax on match (it's called `matches!`) (but follows `if let` chain syntax but without the `= c` after the pattern)

**Comment by Voultapher** [comment]
I agree that using something like `matches!(Ok(x) | Err(x) && x.len() < 100)` causes more issues than it helps. We should definitely not invent a new pattern syntax. Maybe there are options that are less confusing than the existing proposal, but don't require a new pattern syntax?

1. Swap the parameter order around. `assert_matches!(Ok(x) | Err(x) if x.len() < 100, c);` that might help make it more clear that the `if` is part of the pattern and not some `assert_matches` specific thing.
2. Expand the existing `assert!` macro, maybe something like `assert!(match c => Ok(x) | Err(x) if x.len() < 100);` or `assert!(match c { Ok(x) | Err(x) if x.len() < 100 });`. That would use practically no new concepts. With the exception of the omitted curly braces, this would be even closer to the existing `match` syntax. No idea if that is technically possible, and I suspect it would be bound to a new Rust edition. grep.app suggests that ~0.1% of assert usages would collide with this.
3. Disallow if guards in `assert_matches` mimicking constructs like if let, where if guards are also not allowed. This would limit confusion risk at the cost of consistency and applicability. grep.app suggests this would affect ~5.7% of the use cases.
4. Similar to 3. disallow the if guard, but allow a branch body e.g. `assert_matches!(c, Ok(x) | Err(x) => assert!(x.len() < 100));`, this would diverge from `matches!` which is probably a bad idea.
5. Similar to 2 but use a new assert name `assert_pattern!(match x { Ok(x) | Err(x) if x.len() < 100 });`.
6. Have an abstraction that allows to capture the semantics of a single pattern as an object, that can be used and passed around. E.g. `matcher!(Ok(x) | Err(x) if x.len() < 100).assert(c);` or `pattern!(Ok(x) | Err(x) if x.len() < 100).assert_matches(c);`. This would probably require a significant amount of changes. That said I've come across a couple of places in Rust where abstracting over the specific match logic would have been really nice.
7. We use a time machine to replace `if` with `where` in the pattern syntax. And also make binding more clear somehow. `assert_matches!(c, (Ok(x) | Err(x)) where x.len() < 100);`

List of all suggestions:

```rust
1. assert_matches!(c, Ok(x) | Err(x) if x.len() < 100);
2. assert_matches!(Ok(x) | Err(x) if x.len() < 100, c);
3. assert!(match c => Ok(x) | Err(x) if x.len() < 100);
3. assert!(match c { Ok(x) | Err(x) if x.len() < 100 });
4. assert_matches!(c, Ok(x) | Err(x) => assert!(x.len() < 100));
5. assert_pattern!(match x { Ok(x) | Err(x) if x.len() < 100 });
6. matcher!(Ok(x) | Err(x) if x.len() < 100).assert(c);
6. pattern!(Ok(x) | Err(x) if x.len() < 100).assert_matches(c);
7. assert_matches!(c, (Ok(x) | Err(x)) where x.len() < 100);
```

That's after roughly an hour of exploring the design space. I believe there are more options.

Of course there is the elephant in the room that any deviation from the assert_matches crate will make adoption more work. However if the criteria is purely based on that, what point is there in discussing design?

**Comment by tmccombs** [comment]
Another option:
```rust
assert_matches!( c { Ok(x) | Err(x) if x.len() < 100 })
```

And for consistentency also allow the following syntax with the `matches!` macro:

```rust
matches!( c { Ok(x) | Err(x) if len(x) < 100 })
```

I don't like 1 because then the order is the opposite of the existing `matches!` macro. Similarly I'm unhappy with 3 or 4, because it is substantially different from how `matches!` works.

**Comment by nnethercote** [comment]
I'm going to add some meta-commentary because I think this conversation has gotten to a point that's not productive.
- There have been 20 comments in the past couple of days.
- Nine of them by @Voultapher, who strongly opposes the syntax.
- Eleven of them by seven other people, none of whom appear to oppose the syntax.

Ultimately it's up to the libs-api team to decide this. I'm sure they will consider the newest arguments put forth.

**Comment by 8573** [comment]
My two cents are that the precedence of `|` vs `if` in patterns is confusing and that the solution would be not to disallow `if` in one use of patterns but to use parentheses to clarify the precedence, with an optional Clippy lint to require the parentheses.

**Comment by pravic** [comment]
Why are you, guys, arguing in the `assert_matches` topic if the syntax is exactly the same as in [`matches!`](https://doc.rust-lang.org/std/macro.matches.html) macro? You should have had this discussion during `matches!` stabilization - here it's too late because `assert_matches` is simply `assert!(matches!(expr), msg)`.

**Comment by Voultapher** [comment]
> Why are you, guys, arguing in the assert_matches topic if the syntax is exactly the same as in [matches!](https://doc.rust-lang.org/std/macro.matches.html) macro?

Please carefully read what I've previously written. I think there is a meaningful difference. Here are the two core sources of confusion:

1. What part of the pattern does the `if` bin to? `Ok(x) | Err(x) if x.len()`, the Ok, the Err or both? That ship as you said has sailed a long time ago, even long before the `matches!` macro.
2. I think the intuitive reading of `assert_matches!(c, Ok(x) if x.len() < 100)` is: "assert that it matches the pattern **if** the condition is met". However that's not the semantics. It's actually "assert that it matches the pattern **and** the condition is met". I argue that is unique to `assert_matches!` because of the logic inversion. I've made that point a couple times now, for example [here](https://github.com/rust-lang/rust/issues/82775#issuecomment-1920704724).

**Comment by nnethercote** [comment]
Everyone, please stop, the same points are just being rehashed.

**Comment by Rigidity** [comment]
3 years and no stabilization for something as simple as this? :frowning:

**Comment by oli-obk** [comment]
Tracking issues have a tendency to become unmanageable. Please open a dedicated new issue and link to this issue for absolutely any topics you want to discuss or have questions about. See https://github.com/rust-lang/compiler-team/issues/739 for details and discussions on this prospective policy.

**Comment by YarekTyshchenko** [comment]
> FWIW, `assert!(matches!(c, Ok(x) | Err(x) if x.len() < 100));` is already valid.
> 
> Also, about this running example: binding a variable in multiple arms and then using it in a condition is pretty unusual.
> 
> If you don't like the fact that `Ok(x) | Err(x) if x.len() < 100` is a valid pattern, I can understand that. It's arguably surprising. But I don't buy the argument that `assert_matches!` is particularly different to `match` or anywhere else a pattern can appear.

I came looking for `assert_eq!` equivalent but with `matches!` funcionality: Assert eq prints left and right side when there's an inequality, and is very useful is tests. the functionaly equivalent `assert!(matches!(...));` does not print a nice error message :(

Come to think of it, maybe what I want is `assert_eq_matches!(...)` ?

**Comment by RmStorm** [comment]
Hello, there are two unchecked boxes in the top message:

  
>  [ ]  Update the `must_use` message on `is_none` (per [Add messages to `Option`'s and `Result`'s `must_use` annotation for `is_*` #62431 (review)](https://github.com/rust-lang/rust/pull/62431#pullrequestreview-258576936))*
>   [ ]  Mention this macro in the documentation of `matches!()`


The first is done since that PR is merged? And the second one also looks done?
https://github.com/rust-lang/rust/blob/master/library/core/src/macros/mod.rs#L457

Does this mean that `assert_matches!` can be stabilized? What does that process look like and what's the timeline? I'm kinda new here 😅

**Comment by Voultapher** [comment]
https://github.com/rust-lang/rust/pull/121732 is the PR (by me) that improved the documentation which also included "Mention this macro in the documentation of matches!()".

**Comment by huntc** [comment]
So are we able to tick those two remaining boxes then? Thanks.

**Comment by Voultapher** [comment]
@tyilo honestly based on [this](https://github.com/rust-lang/rust/pull/120234#issuecomment-1966986337) comment a year ago where the libs team voted yes to stabilize this, with the only requirement being that an appropriate module would be found to put it in, I had assumed it was stabilized already. When writing that PR it never occurred to me to check that it was actually stabilized.

I agree the link to `assert_matches` should only be made once it's stabilized. The PR mainly focused on the confusing behavior I complained about, and that part is fine and independent of stabilization.

**Comment by tmccombs** [comment]
It has the finished-final-comment-period label. Do we just need a PR to stabilize it? If so, I could make one.

**Comment by safinaskar** [comment]
I think https://github.com/rust-lang/rfcs/pull/3573 fully subsumes `assert_matches`. And https://github.com/rust-lang/rfcs/pull/3573 is more general! Instead of `assert_matches!(x, Some(_))` we can write `assert!(x is Some(_))`, which is shorter and nicer.

So, I think https://github.com/rust-lang/rfcs/pull/3573 should be implemented and then `assert_matches!` should be removed

**Comment by ssokolow** [comment]
RFC 3573 is controversial, as you can see from the nature of the opposition to the idea in the discussion... though given that "detracts from Rust's features being orthogonal and well-factored now that the v1.0 stability promise is blocking removing something else" is one of the complaints, I suppose it *is* the responsible thing to ask to block stabilizing this until a ruling is made on that.

**Comment by nsunderland1** [comment]
Wouldn't `assert!(x is Some(_))` have the same limitations `assert!(matches!(x, Some(_)))` in that it can't give you a `Debug` printout?

```
thread 'main' panicked at src/main.rs:3:5:
assertion failed: matches!(x, Some(_))
```

**Comment by teohhanhui** [comment]
> same limitations as assert!(matches!(x, Some(_))) in that it can't give you a Debug printout?

There was https://github.com/rust-lang/rust/pull/110382

It'd seem [`generic_assert` (RFC 2011)](https://github.com/rust-lang/rust/issues/44838) and `assert_matches` are at odds with one another? At least in terms of the message it's sending to the users: 

1. Use `assert!(...)` for everything. 
2. But use `assert_matches` instead of `assert!(matches!(...))` 🤔

**Comment by tmccombs** [comment]
I wonder if it would be better for this to be part of the generic asset functionality. So you could write something like:

```rust
assert!(matches!(x, Some(_)))
```
and get a useful error message.  Or maybe some assert-specific syntax like:
```rust
assert!(match x => Some(_))
```

or 
```rust
assert!(let Some(_) = x)
```

**Comment by Amanieu** [comment]
Copying my comment from https://github.com/rust-lang/rust/pull/137487#issuecomment-2770801974:

> I don't think a `macros` module is the right way to go around this. What we really want is for `assert_matches` to live in the same place as all the other macros in the standard library, which is currently the crate root (macros are imported with `use std::assert;`).
> 
> Fundamentally the breakage comes from the fact that `assert_matches` is in the _prelude_ and the way macro name resolution works means that we can't prioritize prelude macros lower than glob-imported macros ([#82775 (comment)](https://github.com/rust-lang/rust/issues/82775#issuecomment-1103168321)). However we can avoid this problem by exporting `assert_matches` from the crate root just like other macros but _not_ including it in the prelude. It can be included in the prelude in a future edition.
> 
> Now, the standard library prelude is imported by injecting this code into each crate:
> 
> ```rust
> #[prelude_import]
> use ::std::prelude::rust_2024::*;
> #[macro_use]
> extern crate std;
> ```
> 
> Instead of importing all macros from `std` with `#[macro_use]`, we can explicitly re-export all macros in the `rust_20xx` prelude and remove `#[macro_use]` from the implicit prelude. This allows new macros to be added to the standard library without being a breaking change, while adding them to the prelude can be deferred to a future edition.

**Comment by eirnym** [comment]
> Copying my comment from https://github.com/rust-lang/rust/pull/137487#issuecomment-2770801974:
> 

I'd love to see beginning of this module with this assert

**Comment by Voultapher** [comment]
@eirnym I'm confused by your comment, did you mean to link a different comment?

**Comment by eirnym** [comment]
@Voultapher currently I see assert_matches is located in assert_matches module, which is quite confusing.

**Comment by Voultapher** [comment]
Putting it in `std::assert_matches` is what we've been working on since April.

**Comment by jhpratt** [comment]
This was destabilized in #154999. Reopening. The reason for destabilization seems to be scoping, which is tracked in #154406.

**Comment by jhpratt** [comment]
Looking at release notes, it appears as though the scoping was backported to beta and will land on stable tomorrow (in 1.96).
